File: resources/android/DialogFunctions.kt
package com.nativephp.dialog

import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeError
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.utils.NativeActionCoordinator
import com.google.android.material.snackbar.Snackbar
import org.json.JSONArray

/**
 * Functions related to native alert dialogs
 * Namespace: "Dialog.*"
 */
object DialogFunctions {

    /**
     * Show a native alert dialog with custom buttons
     * Parameters:
     *   - title: (optional) string - Alert title
     *   - message: (optional) string - Alert message body
     *   - buttons: (optional) array of strings - Button titles (defaults to ["OK"])
     *   - id: (optional) string - Custom ID included in event payload
     *   - event: (optional) string - Custom event class name (defaults to "Native\Mobile\Events\Alert\ButtonPressed")
     * Events:
     *   - Fires the specified event (or default) when a button is tapped
     */
    class Alert(private val context: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val title = parameters["title"] as? String
            val message = parameters["message"] as? String
            val id = parameters["id"] as? String
            val event = parameters["event"] as? String ?: "Native\\Mobile\\Events\\Alert\\ButtonPressed"

            // Parse buttons array
            val buttons = mutableListOf<String>()
            when (val buttonsParam = parameters["buttons"]) {
                is JSONArray -> {
                    for (i in 0 until buttonsParam.length()) {
                        buttonsParam.optString(i)?.let { buttons.add(it) }
                    }
                }
                is List<*> -> {
                    buttonsParam.filterIsInstance<String>().forEach { buttons.add(it) }
                }
                is Array<*> -> {
                    buttonsParam.filterIsInstance<String>().forEach { buttons.add(it) }
                }
            }

            if (buttons.isEmpty()) {
                buttons.add("OK")
            }

            // Launch alert on UI thread
            Handler(Looper.getMainLooper()).post {
                try {
                    val coord = NativeActionCoordinator.install(context)
                    coord.launchAlert(
                        title ?: "",
                        message ?: "",
                        buttons.toTypedArray(),
                        id,
                        event
                    )
                } catch (e: Exception) {
                    Log.e("DialogFunctions.Alert", "Error launching alert: ${e.message}", e)
                }
            }

            return emptyMap()
        }
    }

    /**
     * Show a toast notification (uses Snackbar on Android for better UX)
     * Parameters:
     *   - message: string - The message to display
     *   - duration: string (optional) - "short" or "long" (default: "long")
     * Returns:
     *   - success: boolean - True if toast shown successfully
     */
    class Toast(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val message = parameters["message"] as? String
                ?: throw BridgeError.InvalidParameters("message is required")

            val durationParam = parameters["duration"] as? String ?: "long"
            val duration = when (durationParam.lowercase()) {
                "short" -> Snackbar.LENGTH_SHORT
                else -> Snackbar.LENGTH_LONG
            }

            Log.d("Dialog.Toast", "Showing toast (Snackbar): $message")

            return try {
                // Show snackbar on main thread
                Handler(Looper.getMainLooper()).post {
                    // Find a view to anchor the Snackbar to
                    val rootView = activity.findViewById<View>(android.R.id.content)
                    if (rootView != null) {
                        val snackbar = Snackbar.make(rootView, message, duration)

                        // Get current IME (keyboard) insets to avoid overlap
                        val windowInsets = ViewCompat.getRootWindowInsets(rootView)
                        val imeInsets = windowInsets?.getInsets(WindowInsetsCompat.Type.ime())?.bottom ?: 0

                        // Standard Material Design bottom nav height is 56dp
                        val bottomNavHeight = (56 * activity.resources.displayMetrics.density).toInt()

                        // Use whichever is greater: bottom nav or keyboard
                        val bottomOffset = maxOf(bottomNavHeight, imeInsets)

                        val view = snackbar.view
                        val params = view.layoutParams as android.view.ViewGroup.MarginLayoutParams
                        params.setMargins(
                            params.leftMargin,
                            params.topMargin,
                            params.rightMargin,
                            params.bottomMargin + bottomOffset
                        )
                        view.layoutParams = params

                        snackbar.show()
                        Log.d("Dialog.Toast", "Snackbar displayed with bottom offset: ${bottomOffset}px")
                    } else {
                        Log.e("Dialog.Toast", "Could not find root view for Snackbar")
                    }
                }

                mapOf("success" to true)
            } catch (e: Exception) {
                Log.e("Dialog.Toast", "Error showing toast: ${e.message}", e)
                throw BridgeError.ExecutionFailed("Failed to show toast: ${e.message}")
            }
        }
    }
}

File: resources/ios/DialogFunctions.swift
import Foundation
import UIKit

// MARK: - Dialog Function Namespace

/// Functions related to native alert dialogs
/// Namespace: "Dialog.*"
enum DialogFunctions {

    // MARK: - Dialog.Alert

    /// Show a native alert dialog with custom buttons
    /// Parameters:
    ///   - title: (optional) string - Alert title
    ///   - message: (optional) string - Alert message body
    ///   - buttons: (optional) array of strings - Button titles (defaults to ["OK"])
    ///   - id: (optional) string - Custom ID included in event payload
    ///   - eventClass: (optional) string - Custom event class name (defaults to "Native\Mobile\Events\Alert\ButtonPressed")
    /// Events:
    ///   - Fires the specified eventClass (or default) when a button is tapped
    class Alert: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let title = parameters["title"] as? String
            let message = parameters["message"] as? String
            let id = parameters["id"] as? String
            let event = parameters["event"] as? String ?? "Native\\Mobile\\Events\\Alert\\ButtonPressed"

            var buttons: [String] = []
            if let buttonsArray = parameters["buttons"] as? [String] {
                buttons = buttonsArray
            } else if let buttonsArray = parameters["buttons"] as? [Any] {
                buttons = buttonsArray.compactMap { $0 as? String }
            }

            if buttons.isEmpty {
                buttons = ["OK"]
            }

            DispatchQueue.main.async {
                guard let scene = UIApplication.shared.connectedScenes
                          .compactMap({ $0 as? UIWindowScene })
                          .first(where: { $0.activationState == .foregroundActive }),
                      let root = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController else {
                    return
                }

                let alert = UIAlertController(title: title,
                                              message: message,
                                              preferredStyle: .alert)

                for (index, label) in buttons.enumerated() {
                    alert.addAction(UIAlertAction(title: label,
                                                  style: .default) { _ in
                        var payload: [String: Any] = ["index": index, "label": label]
                        if let id = id {
                            payload["id"] = id
                        }

                        LaravelBridge.shared.send?(event, payload)
                    })
                }

                root.present(alert, animated: true)
            }

            return [:]
        }
    }

    // MARK: - Dialog.Toast

    /// Show a toast notification
    /// Parameters:
    ///   - message: string - The message to display
    ///   - duration: string (optional) - "short" or "long" (default: "long")
    /// Returns:
    ///   - success: boolean - True if toast shown successfully
    class Toast: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let message = parameters["message"] as? String else {
                throw BridgeError.invalidParameters("message is required")
            }

            let durationParam = parameters["duration"] as? String ?? "long"
            let duration: TimeInterval = durationParam.lowercased() == "short" ? 2.0 : 4.0

            print("Dialog.Toast called with message: \(message)")

            // Show toast on main thread
            DispatchQueue.main.async {
                ToastManager.shared.show(message: message, duration: duration)
                print("Toast displayed")
            }

            return ["success": true]
        }
    }
}